developer_docs/type_inference.md

```diff
@@ -43,3 +43,33 @@ These objects contain all the information necessary to create a Variable from a
+-   `PrimitiveType` : Representing the category of datatype (integer/floating point/etc)
+-   `PyccelType` : Representing the actual type of the object in Python
+
+Subclasses of `PyccelType` generally fall into one of two categories:
+-   `FixedSizeType`
+-   `ContainerType`
+
+The types can be compared using either the `is` operator or the `==` operator. These operators have slightly different behaviour. All instances of `PyccelType` are singletons so the `is` operator tests if the types are identical. However the `==` operator tests if the types are compatible. For example `PythonNativeFloat() == NumpyFloat64Type()` will return true. This operator should therefore be used when permissive behaviour is required (e.g. when adding elements to a list of `PythonNativeFloat()` we are capable of adding an instance with the type `NumpyFloat64Type()` even if this would not be strictly homogeneous in Python).
+
+#### Fixed Size Type
+A `FixedSizeType` is an object whose size in memory is known and cannot change from one instance to another (e.g. `float64`, `int32`, `void`). In most cases the developer will need the sub-class `FixedSizeNumericType` which refers to the subset of fixed size types which contain numeric values. These objects are characterised by a `primitive_type` describing the category of datatype (integer/floating point/etc) and a `precision`. They additionally implement two magic methods:
+-   `__add__` (+)
+-   `__and__` (&)
+
+The add operator describes what happens when two numeric types are combined in an arithmetic operator. In almost all cases this is sufficient to describe all resulting datatypes. Special cases (e.g. float for integer division) are handled in the associated operator in `ast.operators` or `ast.bitwise_operators`.
+
+The and operator describes what happens when two numeric types are combined in a bitwise comparison operator. This only applies to integers and booleans.
+
+When using these operators on an unknown number of arguments it can be useful to use `NativeGeneric()` as a starting point for the sum.
+
+#### Container Type
+A `ContainerType` is an object which is comprised of `FixedSizeType` objects (e.g. `ndarray`,`list`,`tuple`, custom class). The sub-class `HomogeneousContainerType` describes containers which contain homogeneous data. These objects are characterised by an `element_type`. The elements of a `HomogeneousContainerType` are instances of `PyccelType`, but they can be either `FixedSizeType`s or `ContainerType`s.
+
+`HomogeneousContainerType`s also contain some utility functions. They implement `primitive_type` and `precision` to get the properties of the internal `FixedSizeType` (even if that type is inside another `HomogeneousContainerType`). They also implement `switch_basic_type` which creates a new `HomogeneousContainerType` which is similar to the current `HomogeneousContainerType`. The only difference is that the `FixedSizeType` is exchanged.
```

Comment:
> They also implement `switch_basic_type` which creates a new `HomogeneousContainerType` which is similar to the current `HomogeneousContainerType`. The only difference is that the `FixedSizeType` is exchanged.

Can you please give an example for this so I can understand the use case?

Reply:
I have added more explanation: 5f760e39
This utility is mediumly useful at the moment but it will be much more useful once the rank and order is specified in the container type.